backend/app/models/project_involvement.rb

```diff
@@ -5,10 +5,14 @@ class ProjectInvolvement < ApplicationRecord
   validates_uniqueness_of :project_id, scope: :project_developer_id
 
   after_create do
-    ProjectDeveloperMailer.added_to_project(project_developer, project).deliver_later if project.published?
+    if project.valid? && project.published? && !project.status_changed? # notification is send only when project status has not changed to published, in such case project itself notifies all users on its own
```

Comment:
Do we need `project.valid?` here? My guess this is because it's after_create and not after_commit. I think it is also possible to check changes in after_commit with `saved_change_to_attribute?`

Reply:
Actually we need it here (and it is one of the reasons I hate mails at callbacks). We assign new PD via `involved_project_developer_ids` at controller and flow is following:
 - immediately after `involved_project_developer_ids` changes, appropriate `ProjectInvolvement ` are created/destroyed
 - `ProjectInvolvement` callbacks are triggered
 - `Project` is validated
 - if something fails, it gets rollbacked (which actually does not help with emails)
 
There are two other approaches, but every has its pros and cons:
 - use nested_attributes --> but we would need extra stuff to API controller and probably to backoffice
 - use specialized service to send all emails after all changes to Project are saved --> it would need to collect deleted and added records
 
Soo for now I have added a lot of tests and went for this solution, which I think is easiest one, but we can theoretically go for different one.

Reply:
ok, it can stay this way, but I think `after_commit` won't have those problems, and the problem with rollback and email being already sent.

Reply:
that is also reason why I use `project.status_changed?`, because changes to project are not saved yet.

Reply:
Oki douky, I will try it @tsubik , if it helps, I will switch to `after_commit` callback ;)

Reply:
But I'm not sure if `after_delete` with `after_commit` on: :destroy you will have that info about `destroyed_by_association`

Reply:
Maybe we cannot run from the service layer and eventually that would be a good choice to add a couple services, at least for projects, to update or delete.

Reply:
Agree with you, at future we should probably move more complex actions to services, sometimes that logic is starting to be quite complex and controller/model stuff is not enough